src/main.rs

```diff
@@ -1,3 +1,203 @@
+use std::io::{self, prelude::*};
+use std::os::unix::net::UnixStream;
+use std::process::{Command, Stdio};
+use std::{thread, time};
+
+#[repr(u8)]
+enum MsgType {
+    MsgQuit = 1,
+    MsgRunProcess,
+}
+
+#[repr(u8)]
+enum MsgSubType {
+    SubMsgEnd = 0,
+    SubMsgRunProcessBin,
+    SubMsgRunProcessArg,
+    SubMsgRunProcessEnv,
+    SubMsgRunProcessUid,
+    SubMsgRunProcessRfd,
+}
+
+#[repr(u8)]
+#[derive(Debug)]
+enum RespType {
+    RespOk = 0,
+    RespErr,
+    NotifyOutputAvailable,
+    NotifyProcessDied,
+}
+
+impl RespType {
+    fn from_u8(x: u8) -> io::Result<Self> {
+        match x {
+            0 => Ok(RespType::RespOk),
+            1 => Ok(RespType::RespErr),
+            2 => Ok(RespType::NotifyOutputAvailable),
+            3 => Ok(RespType::NotifyProcessDied),
+            _ => Err(io::Error::new(
+                io::ErrorKind::InvalidData,
+                "Invalid response type",
+            )),
+        }
+    }
+}
+
+struct GuestAgent {
+    stream: UnixStream,
+    last_msg_id: u64,
+}
+
+impl GuestAgent {
+    fn connected(path: &str) -> io::Result<GuestAgent> {
+        loop {
+            match UnixStream::connect(path) {
+                Ok(s) => {
+                    break Ok(GuestAgent {
+                        stream: s,
+                        last_msg_id: 0,
+                    })
+                }
+                Err(err) => match err.kind() {
+                    io::ErrorKind::NotFound => {
+                        println!("Waiting for Guest Agent socket ...");
+                        thread::sleep(time::Duration::from_secs(1));
+                    }
+                    _ => break Err(err),
+                },
+            };
+        }
+    }
+
+    fn get_new_msg_id(&mut self) -> u64 {
+        self.last_msg_id += 1;
+        self.last_msg_id
+    }
+
+    fn create_header(msg_id: u64, typ: MsgType) -> Vec<u8> {
```

Comment:
`typ` -> `type`

Reply:
`type` is a keyword, so it cannot be used as identifier.